labkey-client-api/src/org/labkey/remoteapi/Command.java

```
@@ -520,60 +512,55 @@ protected URI getActionUrl(Connection connection, String folderPath) throws URIS
-     * @return The query string
-     * @throws CommandException Thrown if there is a problem encoding the query string parameter names or values
+     * @return The URI with parameters added
+     * @throws CommandException Thrown if there is a problem building the URI
      */
-    protected String getQueryString() throws CommandException
+    protected URI addParameters(URI uri) throws CommandException, URISyntaxException
     {
         Map<String, Object> params = getParameters();
 
         //add the required version if it's > 0
         if (getRequiredVersion() > 0)
             params.put(CommonParameters.apiVersion.name(), getRequiredVersion());
 
-        StringBuilder qstring = new StringBuilder();
-        URLCodec urlCodec = new URLCodec();
-        try
+        if (params.isEmpty())
+            return uri;
+
+        URIBuilder builder = new URIBuilder(uri);
+
+        for (String name : params.keySet())
         {
-            for (String name : params.keySet())
+            Object value = params.get(name);
+            if (value instanceof Collection<?> col)
             {
-                Object value = params.get(name);
-                if (value instanceof Collection)
+                for (Object o : col)
                 {
-                    for (Object o : ((Collection) value))
-                    {
-                        appendParameter(qstring, urlCodec, name, o);
-                    }
-                }
-                else
-                {
-                    appendParameter(qstring, urlCodec, name, value);
+                    addParameter(builder, name, o);
                 }
             }
+            else
+            {
+                addParameter(builder, name, value);
+            }
+        }
+
+        try
+        {
+            return builder.build();
         }
-        catch(EncoderException e)
+        catch (URISyntaxException e)
         {
             throw new CommandException(e.getMessage());
         }
```

`setParameters` is declared as throwing `URISyntaxException`, why not just let this bubble up instead of wrapping in a `CommandException`?
Also, I would argue that `Command._execute` shouldn't wrap it in a `CommandException` either. Wrapping in an `IllegalArgumentException` seems more appropriate to me since the most likely cause of such an exception is somebody specifying bad query parameters or some such thing.
Although, I suppose, throwing unchecked exceptions from an API is probably bad practice.

---

I was just following the existing pattern, which I agree is inconsistent. As I think I noted somewhere, I'm planning to merge the create URL and add parameters steps into a single method at some point. Will look at reconciling exception handling at that point. But I think this set of PRs is close enough and I need to move on...